### ISSUE TITLE
[java] SwitchStmtsShouldHaveDefault false positive with pattern matching

### ISSUE THREAD

**Issue by emouty** [pmd/pmd#4813]
**Affects PMD Version:** 7.0.0-rc4 

**Rule:**

Please provide the rule name and a link to the rule documentation:
- https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#switchstmtsshouldhavedefault
- https://docs.pmd-code.org/latest/pmd_rules_java_performance.html#toofewbranchesforaswitchstatement

**Description:**

**Code Sample demonstrating the issue:**

```java
    public sealed interface AcceptableResult permits Success, Warning {
        public String message();
    }

    public final class Success implements AcceptableResult {

        @Override
        public String message() {
            return "Success!";
        }
    }

    public abstract class Failure {

        abstract public String message();
    }

    public non-sealed class Warning extends Failure implements AcceptableResult {

        @Override
        public String message() {
            return "Oopsie";
        }
    }

    public class ProviderWarning extends Warning {

        @Override
        public String message() {
            return "Ohoh";
        }
    }

    public void test(AcceptableResult result) {
        switch (result) {
            case Warning failure -> System.out.println("WARNING " + failure.message());
            case Success success -> System.out.println(success.message());
        }
    }
```

**Expected outcome:**

PMD reports violations on the switch SwitchStmtsShouldHaveDefault and TooFewBranchesForASwitchStatement. For SwitchStmtsShouldHaveDefault it's a false positive since we have covered all of the cases here. As for TooFewBranchesForASwitchStatement I think it's also a false positive in this case, since using a pattern matching switch is the cleanest way to do this kind of stuff even if there are only 2 branches.  

**Running PMD through:** * Maven*

**Comment by cowwoc** [comment]
Just to confirm, I am also seeing the false-positive for SwitchStmtsShouldHaveDefault for switch statements that cover all of the cases, so there is definitely a bug there.
